File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/controller/ImagesRestController.java
package com.raxim.myscoutee.profile.controller;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.util.Base64;
import java.util.UUID;

import org.springframework.data.util.Pair;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.raxim.myscoutee.common.util.FileUtil;
import com.raxim.myscoutee.profile.data.dto.rest.ImageDTO;

import jakarta.servlet.http.HttpServletResponse;

@RestController
@RequestMapping("images")
public class ImagesRestController {

    @GetMapping("/{name}")
    public void download(@PathVariable String name, HttpServletResponse response) throws IOException {
        Pair<String, String> tempPath = FileUtil.tempToPath(name, false);
        String separator = tempPath.getFirst();
        String tmpDir = tempPath.getSecond();

        File file = new File(tmpDir + separator + "_" + name);
        FileUtil.copyFileToHttp(response, file);
    }

    @PostMapping(value = { "", "/{name}" })
    public ResponseEntity<ImageDTO> upload(
            @PathVariable(required = false) String name,
            @RequestParam(value= "orig", required = false) MultipartFile orig,
            @RequestParam("edited") MultipartFile edited) {
        String fileName = name != null ? name : UUID.randomUUID().toString();
        try {
            Pair<String, String> tempPath = FileUtil.tempToPath(fileName, true);
            String separator = tempPath.getFirst();
            String tmpDir = tempPath.getSecond();

            if (edited != null && !edited.isEmpty()) {
                String completeImageData = new String(edited.getBytes(), Charset.forName("UTF-8"));
                String imageDataBytes = completeImageData.substring(completeImageData.indexOf(",") + 1);
                InputStream imgStream = new ByteArrayInputStream(Base64.getMimeDecoder().decode(imageDataBytes));

                FileUtil.save(imgStream, tmpDir + separator + "_" + fileName);
            }

            if (orig != null && !orig.isEmpty()) {
                InputStream origIs = new ByteArrayInputStream(orig.getBytes());
                String origFile = tmpDir + separator + "_" + fileName + "_orig";

                FileUtil.save(origIs, origFile);

                return ResponseEntity.status(HttpStatus.CREATED).body(new ImageDTO(fileName));
            } else {
                Pair<String, String> fullDir = FileUtil.uuidToPath(fileName, true);
                String fullDirName = fullDir.getSecond();
                String origFileName = fullDirName + separator + "_" + fileName + "_orig";
                String tmpOrigFileName = tmpDir + separator + "_" + fileName + "_orig";

                File permOrigFile = new File(origFileName);
                File tmpOrigFile = new File(tmpOrigFileName);

                if (permOrigFile.exists() || tmpOrigFile.exists()) {
                    return ResponseEntity.status(HttpStatus.OK).body(new ImageDTO(fileName));
                } else {
                    return ResponseEntity.badRequest().build();
                }
            }
        } catch (Exception e) {
            return ResponseEntity.badRequest().build();
        }
    }
}

File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/controller/ImagesRestController.kt
package com.raxim.myscoutee.profile.controller

import com.raxim.myscoutee.common.FileUtil
import com.raxim.myscoutee.common.save
import com.raxim.myscoutee.profile.data.dto.rest.Image
import org.bson.BsonBinary
import org.bson.BsonBinarySubType
import org.bson.UuidRepresentation
import org.springframework.data.rest.webmvc.RepositoryRestController
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.*
import java.nio.charset.Charset
import java.util.*
import javax.servlet.http.HttpServletResponse

fun String.asUUID(): String {
    val bytes = Base64.getDecoder().decode(this)
    return BsonBinary(BsonBinarySubType.UUID_LEGACY, bytes).asUuid(UuidRepresentation.JAVA_LEGACY).toString()
}

@RepositoryRestController
@RequestMapping("images")
class ImagesRestController {

    @Throws(
        IOException::class
    )
    @GetMapping("/{name}")
    fun download(@PathVariable name: String, response: HttpServletResponse) {
        val (separator, tmpDir) = FileUtil.tempToPath(name)

        val file = File("$tmpDir${separator}_$name")
        FileUtil.copyFileToHttp(response, file)
    }

    // images under user uid folder
    @PostMapping(value = ["", "/{name}"])
    fun upload(
        @PathVariable name: String? = null,
        @RequestParam orig: MultipartFile? = null, @RequestParam edited: MultipartFile? = null
    ): ResponseEntity<Image> {

        val fileName: String = name ?: UUID.randomUUID().toString()
        try {
            val (separator, tmpDir) = FileUtil.tempToPath(fileName, true)

            if (edited != null && !edited.isEmpty) {
                val completeImageData = String(edited.bytes, Charset.forName("UTF-8"))
                val imageDataBytes = completeImageData.substring(completeImageData.indexOf(",") + 1)
                val imgStream: InputStream = ByteArrayInputStream(Base64.getMimeDecoder().decode(imageDataBytes))

                imgStream.save("$tmpDir${separator}_$fileName")
            }

            return if (orig != null && !orig.isEmpty) {
                val origIs = ByteArrayInputStream(orig!!.bytes)
                val origFile = "$tmpDir${separator}_${fileName}_orig"

                origIs.save(origFile)

                ResponseEntity(Image(fileName), HttpStatus.CREATED);
            } else {
                val (_, fullDir) = FileUtil.uuidToPath(fileName, true)
                val origFileName = "$fullDir${separator}_${fileName}_orig"
                val tmpOrigFileName = "$tmpDir${separator}_${fileName}_orig"

                val permOrigFile = File(origFileName)
                val tmpOrigFile = File(tmpOrigFileName)

                if (permOrigFile.exists() || tmpOrigFile.exists()) {
                    ResponseEntity(Image(fileName), HttpStatus.OK);
                } else {
                    ResponseEntity(HttpStatus.BAD_REQUEST)
                }
            }
        } catch (e: Exception) {
            return ResponseEntity.badRequest().build()
        }
    }
}